### PR TITLE
Casper krs/product dynamic pricing

### PR DESCRIPTION
**Pull request by casperlotter**
Update existing dynamic pricing components and apply for product context.

Zeplin:
Screenshot 2023-05-18 152844

Implementation:

Screenshot 2023-05-18 152641

A note here @matthampton-stock2shop 

There are some things I will add in a future PR:

1. ProductMaster type of component as we have done in products.
2. VersionControl as well.

There is a lot of foundation types and components done for this on another branch by Abigail and I will do the rest of that conversion once it is in so that we do not duplicate effort.

### CODE REVIEW

File: src/components/PricingContracts/AddPricingContractModal.svelte [hunk 1]
```diff
@@ -10,14 +10,18 @@
...
+  import IconCustomer from '../Icons/IconCustomer.svelte';
   import IconPricing from '../Icons/IconPricing.svelte';
 
   export let open: boolean;
-  const index = ElasticSearchIndex.PRODUCTS;
-
   export let addContract: (contract: PricingContract) => void;
   export let contract: Writable<PricingContract>;
 
+  let index = ElasticSearchIndex.CUSTOMERS;
```
Comment (justinengle-krs) [inline src/components/PricingContracts/AddPricingContractModal.svelte:20]: this is maybe a bit of a personal preference but I prefer something like this as a ternary. A bit more 'declarative' in a sense. But I dont have a strong reason to change it.

`let index =  $contract.forCustomer() ? ElasticSearchIndex.CUSTOMERS : ElasticSearchIndex.PRODUCTS;`

File: src/components/PricingContracts/DynamicPricing.svelte [hunk 2]
```diff
@@ -1,36 +1,41 @@
...
-  export let customerMetaRepo: CustomerMetaRepo;
+  export let customerId = null;
+  export let productId = null;
+  export let metaRepo: CustomerMetaRepo | ProductMetaRepo;
   export let invalidateCache: () => void;
   export let inlineNotification: Writable<EventNotification | null>;
   export let addContractModel: Writable<PricingContract>;
   export let open: boolean = false;
 
+  let entity = customerId ? 'product' : 'customer';
```
Comment (justinengle-krs) [inline src/components/PricingContracts/DynamicPricing.svelte:21]: does this need to be a VM property? or can it by scoped to where it's used

File: src/repos/product-meta-repo.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
+import { post, del, put, get } from '@/lib/api/api';
+import type { APIRequest } from '@/lib/api/types';
+import type { ApiResult } from '../models/api-result';
+import type { ProductMeta } from '../models/meta';
+
+interface SystemProductMeta {
```
Comment (justinengle-krs) [inline src/repos/product-meta-repo.ts:6]: just to keep the standards the same, can these interfaces sit in our model folder? and repo is just the api calls

**Review by justinengle-krs — DISMISSED** [review]

File: src/stories/routes/25-ProductContract.stories.svelte [hunk 4]
```diff
@@ -1,15 +1,29 @@
 <script lang="ts">
-  // eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (matthampton-stock2shop) [inline src/stories/routes/25-ProductContract.stories.svelte:2]: https://www.chromatic.com/build?appId=63ef6cddab1dec384359e1cd&number=96

I did a chromatic run and made a couple of comments

File: src/models/meta.ts [hunk 5]
```diff
@@ -13,5 +13,9 @@ export interface Meta {
 }
 
 export interface CustomerMeta extends Meta {
-  customer_id: string;
+  customer_id?: string;
```
Comment (matthampton-stock2shop) [inline src/models/meta.ts:16]: Is this optional?

File: src/models/meta.ts [hunk 6]
```diff
@@ -13,5 +13,9 @@ export interface Meta {
...
+}
+
+export interface ProductMeta extends Meta {
+  product_id?: string;
```
Comment (matthampton-stock2shop) [inline src/models/meta.ts:20]: And this?

**Review by matthampton-stock2shop — CHANGES_REQUESTED** [review]

**Review by matthampton-stock2shop — APPROVED** [review]

**Pull request merged**